### PR TITLE
feat: Apply Signature Verification for SellData and Build Tx Command-Line

### PR DESCRIPTION
**Pull request by inchori**
What I done
- Build Tx Command-Line for ```SellData```.
- Apply Verification for DataValidator Signature for ```SellData```.
- Register ```MsgSellData``` in ```types/codec.go```.

TODO
- Replace ```max_num_data``` field and ```cur_num_data``` field with pricePerData.

### CODE REVIEW

File: x/market/client/cli/txDeal.go [hunk 1]
```diff
@@ -82,3 +109,48 @@ func parseCreateDealFlags(fs *flag.FlagSet) (*createDealInputs, error) {
...
+		DataValidatorAddress: sellData.Cert.UnsignedCert.DataValidatorAddress,
+		RequesterAddress:     sellData.Cert.UnsignedCert.RequesterAddress,
+	}
+
+	signed := types.DataValidationCertificate{
+		UnsignedCert: &unSigned,
+		Signature:    []byte(sellData.Cert.Signature),
+	}
+
+	msg := types.NewMsgSellData(signed, clientCtx.GetFromAddress().String())
```
Comment (gyuguen) [inline x/market/client/cli/txDeal.go:132]: If you accept the contents of the [previous PR](https://github.com/medibloc/panacea-core/pull/255/files#r788414624), this part should also be modified.

File: x/market/keeper/deal.go [hunk 2]
```diff
@@ -198,3 +198,19 @@ func (k Keeper) SetCurNumData(deal types.Deal) {
...
+		return false, sdkerrors.Wrapf(err, "invalid marshaled value")
+	}
+
+	isValid := validatorAcc.GetPubKey().VerifySignature(unSignedMarshaled, cert.GetSignature())
+	if !isValid {
+		return false, sdkerrors.Wrapf(types.ErrInvalidSignature, "%s", cert.GetSignature())
+	}
+
+	return isValid, nil
+}
```
Comment (gyuguen) [inline x/market/keeper/deal.go:216]:
```suggestion
func (k Keeper) Verify(ctx sdk.Context, validatorAddr sdk.AccAddress, unSignedCert types.UnsignedDataValidationCertificate, cert types.DataValidationCertificate) error {
	validatorAcc := k.accountKeeper.GetAccount(ctx, validatorAddr)

	unSignedMarshaled, err := unSignedCert.Marshal()
	if err != nil {
		return sdkerrors.Wrapf(err, "invalid marshaled value")
	}

	isValid := validatorAcc.GetPubKey().VerifySignature(unSignedMarshaled, cert.GetSignature())
	if !isValid {
		return sdkerrors.Wrapf(types.ErrInvalidSignature, "%s", cert.GetSignature())
	}

	return nil
}
```
I think the `bool` value and `error` play a similar role.
How about removing the bool?

**Review by gyuguen — COMMENTED** [review]
Thank you. I left the few comment. 😀

File: x/market/keeper/deal.go [hunk 3]
```diff
@@ -135,7 +135,7 @@ func (k Keeper) SellOwnData(ctx sdk.Context, seller sdk.AccAddress, cert types.D
 
 	// Is deal activated
 	if findDeal.GetStatus() != ACTIVE {
-		return sdk.Coin{}, fmt.Errorf("deal is not activated")
+		return sdk.Coin{}, sdkerrors.Wrap(err, "deal is not activated")
```
Comment (hansol-medi) [inline x/market/keeper/deal.go:138]: This `err` is an error from `k.GetDeal`.
In keeper func, why don't you define sentinel error and use it? If then, I think you don't have to wrap the sentinel error here.

File: x/market/keeper/deal.go [hunk 4]
```diff
@@ -155,7 +155,7 @@ func (k Keeper) SellOwnData(ctx sdk.Context, seller sdk.AccAddress, cert types.D
 		}
 	}
 	if !flag {
-		return sdk.Coin{}, fmt.Errorf("data validator is invalid address")
+		return sdk.Coin{}, sdkerrors.Wrap(err, "data validator is invalid address")
```
Comment (hansol-medi) [inline x/market/keeper/deal.go:158]: Same here.

File: x/market/keeper/deal_test.go [hunk 5]
```diff
@@ -130,11 +132,23 @@ func (suite *dealTestSuite) TestSellOwnData() {
...
+//
+//	validatorAddr, err := sdk.AccAddressFromBech32(cert.UnsignedCert.GetDataValidatorAddress())
+//	suite.Require().NoError(err)
+//	suite.Require().Equal(newAddr, validatorAddr)
+//
+//	verify, err := suite.MarketKeeper.Verify(suite.Ctx, validatorAddr, *cert.UnsignedCert, cert)
+//	suite.Require().Equal(true, verify)
+//	suite.Require().NoError(err)
+//}
+
```
Comment (inchori) [inline x/market/keeper/deal_test.go:151]: This method is related to issue #259 .

File: x/market/keeper/deal.go [hunk 6]
```diff
@@ -198,3 +198,19 @@ func (k Keeper) SetCurNumData(deal types.Deal) {
 	curNumData := deal.GetCurNumData() + 1
 	deal.CurNumData = curNumData
 }
+
+func (k Keeper) Verify(ctx sdk.Context, validatorAddr sdk.AccAddress, unSignedCert types.UnsignedDataValidationCertificate, cert types.DataValidationCertificate) (bool, error) {
```
Comment (hansol-medi) [inline x/market/keeper/deal.go:202]: The `cert` has `unsignedCertificate` and `signature`, so you don't have to the `unSignedCert` argument.

**Review by hansol-medi — COMMENTED** [review]
Thank you for your works!

**Pull request merged**